File: app/src/main/java/vn/loitp/app/activity/animation/confetti/FallingConfettiFromPointActivity.java
package vn.loitp.app.activity.animation.confetti;

import android.content.res.Resources;
import android.os.Bundle;
import android.view.View;

import java.util.ArrayList;
import java.util.List;

import loitp.basemaster.R;
import vn.loitp.core.base.BaseFontActivity;
import vn.loitp.views.animation.confetti.CommonConfetti;
import vn.loitp.views.animation.confetti.ConfettiManager;
import vn.loitp.views.animation.confetti.ConfettiSource;

//https://github.com/jinatonic/confetti
public class FallingConfettiFromPointActivity extends BaseFontActivity implements View.OnClickListener {
    protected int goldDark, goldMed, gold, goldLight;
    protected int[] colors;

    private final List<ConfettiManager> activeConfettiManagers = new ArrayList<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        findViewById(R.id.generate_confetti_once_btn).setOnClickListener(this);
        findViewById(R.id.generate_confetti_stream_btn).setOnClickListener(this);
        findViewById(R.id.generate_confetti_infinite_btn).setOnClickListener(this);
        findViewById(R.id.generate_confetti_stop_btn).setOnClickListener(this);

        final Resources res = getResources();
        goldDark = res.getColor(R.color.gold_dark);
        goldMed = res.getColor(R.color.gold_med);
        gold = res.getColor(R.color.gold);
        goldLight = res.getColor(R.color.gold_light);
        colors = new int[]{goldDark, goldMed, gold, goldLight};
    }

    @Override
    protected boolean setFullScreen() {
        return false;
    }

    @Override
    protected String setTag() {
        return getClass().getSimpleName();
    }

    @Override
    protected int setLayoutResourceId() {
        return R.layout.activity_falling_confetti_from_top;
    }

    @Override
    public void onClick(View v) {
        final int id = v.getId();
        if (id == R.id.generate_confetti_once_btn) {
            activeConfettiManagers.add(generateOnce());
        } else if (id == R.id.generate_confetti_stream_btn) {
            activeConfettiManagers.add(generateStream());
        } else if (id == R.id.generate_confetti_infinite_btn) {
            activeConfettiManagers.add(generateInfinite());
        } else {
            for (ConfettiManager confettiManager : activeConfettiManagers) {
                confettiManager.terminate();
            }
            activeConfettiManagers.clear();
        }
    }

    protected ConfettiManager generateOnce() {
        return getCommonConfetti().oneShot();
    }

    protected ConfettiManager generateStream() {
        return getCommonConfetti().stream(3000);
    }

    protected ConfettiManager generateInfinite() {
        return getCommonConfetti().infinite();
    }

    private CommonConfetti getCommonConfetti() {
        final int size = getResources().getDimensionPixelSize(R.dimen.default_confetti_size);
        final ConfettiSource confettiSource = new ConfettiSource(-size, -size);
        final CommonConfetti commonConfetti =
                CommonConfetti.rainingConfetti(getRootView(), confettiSource, colors);

        final Resources res = getResources();
        final int velocitySlow = res.getDimensionPixelOffset(R.dimen.default_velocity_slow);
        final int velocityNormal = res.getDimensionPixelOffset(R.dimen.default_velocity_normal);
        final int velocityFast = res.getDimensionPixelOffset(R.dimen.default_velocity_fast);

        // Further configure it
        commonConfetti.getConfettiManager()
                .setVelocityX(velocityFast, velocityNormal)
                .setAccelerationX(-velocityNormal, velocitySlow)
                .setTargetVelocityX(0, velocitySlow / 2)
                .setVelocityY(velocityNormal, velocitySlow);

        return commonConfetti;
    }
}

File: app/src/main/java/vn/loitp/app/activity/animation/confetti/FallingConfettiFromPointActivity.kt
package vn.loitp.app.activity.animation.confetti

import android.os.Bundle
import android.view.View
import androidx.core.content.ContextCompat
import com.core.base.BaseFontActivity
import com.views.animation.confetti.CommonConfetti
import com.views.animation.confetti.ConfettiManager
import com.views.animation.confetti.ConfettiSource
import kotlinx.android.synthetic.main.activity_falling_confetti_from_top.*
import vn.loitp.app.R
import java.util.*

//https://github.com/jinatonic/confetti
class FallingConfettiFromPointActivity : BaseFontActivity(), View.OnClickListener {
    private var goldDark: Int = 0
    private var goldMed: Int = 0
    private var gold: Int = 0
    private var goldLight: Int = 0
    private lateinit var colors: IntArray

    private val activeConfettiManagers = ArrayList<ConfettiManager>()

    private// Further configure it
    val commonConfetti: CommonConfetti
        get() {
            val size = resources.getDimensionPixelSize(R.dimen.default_confetti_size)
            val confettiSource = ConfettiSource(-size, -size)
            val commonConfetti = CommonConfetti.rainingConfetti(rootView, confettiSource, colors)

            val res = resources
            val velocitySlow = res.getDimensionPixelOffset(R.dimen.default_velocity_slow)
            val velocityNormal = res.getDimensionPixelOffset(R.dimen.default_velocity_normal)
            val velocityFast = res.getDimensionPixelOffset(R.dimen.default_velocity_fast)
            commonConfetti.confettiManager
                    .setVelocityX(velocityFast.toFloat(), velocityNormal.toFloat())
                    .setAccelerationX((-velocityNormal).toFloat(), velocitySlow.toFloat())
                    .setTargetVelocityX(0f, (velocitySlow / 2).toFloat())
                    .setVelocityY(velocityNormal.toFloat(), velocitySlow.toFloat())

            return commonConfetti
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        findViewById<View>(R.id.generate_confetti_once_btn).setOnClickListener(this)
        findViewById<View>(R.id.generate_confetti_stream_btn).setOnClickListener(this)
        findViewById<View>(R.id.generate_confetti_infinite_btn).setOnClickListener(this)
        findViewById<View>(R.id.generate_confetti_stop_btn).setOnClickListener(this)

        goldDark = ContextCompat.getColor(activity, R.color.gold_dark)
        goldMed = ContextCompat.getColor(activity, R.color.gold_med)
        gold = ContextCompat.getColor(activity, R.color.gold)
        goldLight = ContextCompat.getColor(activity, R.color.gold_light)
        colors = intArrayOf(goldDark, goldMed, gold, goldLight)
    }

    override fun setFullScreen(): Boolean {
        return false
    }

    override fun setTag(): String {
        return javaClass.simpleName
    }

    override fun setLayoutResourceId(): Int {
        return R.layout.activity_falling_confetti_from_top
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.generate_confetti_once_btn -> activeConfettiManagers.add(generateOnce())
            R.id.generate_confetti_stream_btn -> activeConfettiManagers.add(generateStream())
            R.id.generate_confetti_infinite_btn -> activeConfettiManagers.add(generateInfinite())
            else -> {
                for (confettiManager in activeConfettiManagers) {
                    confettiManager.terminate()
                }
                activeConfettiManagers.clear()
            }
        }
    }

    private fun generateOnce(): ConfettiManager {
        return commonConfetti.oneShot()
    }

    private fun generateStream(): ConfettiManager {
        return commonConfetti.stream(3000)
    }

    private fun generateInfinite(): ConfettiManager {
        return commonConfetti.infinite()
    }
}
